stl/inc/type_traits

```diff
@@ -521,6 +433,110 @@ _INLINE_VAR constexpr bool is_volatile_v<volatile _Ty> = true;
+    };
+
+_CLASS_DEFINE_CV_REF_NOEXCEPT(_IS_FUNCTION_ELLIPSIS)
+#undef _IS_FUNCTION_ELLIPSIS
+
+template <class _Ty>
+_INLINE_VAR constexpr bool is_function_v =
+    !is_const_v<const _Ty> && !is_reference_v<_Ty>; // determine whether _Ty is a function
+
+template <class _Ty>
+struct is_function : bool_constant<is_function_v<_Ty>> {};
+
+template <class _Ty>
+struct _Is_memfunptr { // base class for member function pointer predicates
+    using _Bool_type = false_type; // NB: members are user-visible via _Weak_types
+};
+
+#define _IS_MEMFUNPTR(CALL_OPT, CV_OPT, REF_OPT, NOEXCEPT_OPT)                            \
+    template <class _Ret, class _Arg0, class... _Types>                                   \
+    struct _Is_memfunptr<_Ret (CALL_OPT _Arg0::*)(_Types...) CV_OPT REF_OPT NOEXCEPT_OPT> \
+        : _Arg_types<CV_OPT _Arg0*, _Types...> {                                          \
+        using _Bool_type = true_type;                                                     \
+        _CXX17_DEPRECATE_ADAPTOR_TYPEDEFS typedef _Ret result_type;                       \
+        using _Class_type = _Arg0;                                                        \
+        using _Guide_type = enable_if<!is_same_v<int REF_OPT, int&&>, _Ret(_Types...)>;   \
+    };
+
+_MEMBER_CALL_CV_REF_NOEXCEPT(_IS_MEMFUNPTR)
+#undef _IS_MEMFUNPTR
+
+#define _IS_MEMFUNPTR_ELLIPSIS(CV_REF_NOEXCEPT_OPT)                          \
+    template <class _Ret, class _Arg0, class... _Types>                      \
+    struct _Is_memfunptr<_Ret (_Arg0::*)(_Types..., ...)                     \
+            CV_REF_NOEXCEPT_OPT> { /* no calling conventions for ellipsis */ \
+        using _Bool_type = true_type;                                        \
+        _CXX17_DEPRECATE_ADAPTOR_TYPEDEFS typedef _Ret result_type;          \
+        using _Class_type = _Arg0;                                           \
+        using _Guide_type = enable_if<false>;                                \
+    };
+
+_CLASS_DEFINE_CV_REF_NOEXCEPT(_IS_MEMFUNPTR_ELLIPSIS)
+#undef _IS_MEMFUNPTR_ELLIPSIS
+
+// STRUCT TEMPLATE is_member_function_pointer
+template <class _Ty>
+struct _Is_member_pointer {
+    static constexpr bool value = false;
+};
+
+template <class _Ty1, class _Ty2>
+struct _Is_member_pointer<_Ty1 _Ty2::*> {
+    static constexpr bool value = true;
+    using _Member_type          = _Ty2;
+};
+
+template <class _Ty>
+_INLINE_VAR constexpr bool is_member_function_pointer_v = // determine whether _Ty is a pointer to member function
+    conjunction_v<is_member_pointer<_Ty>, is_function<typename _Is_member_pointer<_Ty>::_Member_type>>;
+
+template <class _Ty>
+struct is_member_function_pointer : bool_constant<is_member_function_pointer_v<_Ty>> {};
+
+
+template <class _Ty, bool _Pmf = is_member_function_pointer_v<_Ty>>
+struct _Is_member_object_pointer { // determine whether _Ty is a pointer to member object

```

Comment:
`_Is_member_object_pointer` is now only used by `invoke`; we can push this definition (and the partial specialization which is weirdly on 535) down above the `_IMPLEMENT_INVOKE` machinery.

Reply:
Done